src/engine/Server.cpp

```diff
@@ -492,6 +494,139 @@ Awaitable<void> Server::process(
+                          transformSparqlTripleComponent(std::move(triple.p_)),
+                          transformSparqlTripleComponent(std::move(triple.o_)),
+                          transformGraph(std::move(triple.g_))};
+      };
+  std::vector<std::array<IdOrVariable, 4>> toInsertTemplates =
+      ad_utility::transform(std::move(graphUpdate.toInsert_),
+                            transformSparqlTripleSimple);
+  std::vector<std::array<IdOrVariable, 4>> toDeleteTemplates =
+      ad_utility::transform(std::move(graphUpdate.toDelete_),
+                            transformSparqlTripleSimple);
+
+  // To calculate the set of quads for the update, for each row in the template
+  // resolve the variables using each row of the result to obtain a quad without
+  // variables. These quads can then be used with `DeltaTriples`.
+  auto resolveVariable = [](const IdTable& idTable, const size_t& row,
+                            const VariableToColumnMap& variableColumns,
+                            IdOrVariable idOrVar) -> std::optional<Id> {
+    if (std::holds_alternative<Variable>(idOrVar)) {
+      auto var = std::get<Variable>(std::move(idOrVar));
+      if (!variableColumns.contains(var)) {
+        return std::nullopt;
+      } else {
+        return idTable(row, variableColumns.at(var).columnIndex_);
+      }
+    } else if (std::holds_alternative<Id>(idOrVar)) {
+      return std::get<Id>(idOrVar);
+    } else {
+      AD_FAIL();
+    }
+  };
+  auto computeQuadsForResultRow = [&resolveVariable](
+                                      auto& templates, auto& result,
+                                      const IdTable& idTable, uint64_t row,
+                                      const VariableToColumnMap&
+                                          variableColumns) {
+    for (const auto& [s, p, o, g] : templates) {
+      auto subject = resolveVariable(idTable, row, variableColumns, s);
+      auto predicate = resolveVariable(idTable, row, variableColumns, p);
+      auto object = resolveVariable(idTable, row, variableColumns, o);
+      auto graph = resolveVariable(idTable, row, variableColumns, g);
+
+      if (!subject.has_value() || !predicate.has_value() ||
+          !object.has_value() || !graph.has_value()) {
+        continue;
+      }
+      result.emplace_back(std::array{*subject, *predicate, *object, *graph});
+    }
+    return result;
+  };
+  std::vector<IdTriple<>> toInsert;
+  std::vector<IdTriple<>> toDelete;
+  // The maximum result size is size(query result) x num template rows. The
+  // actual result can be smaller if there are template rows with variables for
+  // which a result row does not have a value.
+  toInsert.reserve(res->idTable().size() * toInsertTemplates.size());
+  toDelete.reserve(res->idTable().size() * toDeleteTemplates.size());
+  for (const auto& [pair, range] :
+       ExportQueryExecutionTrees::getRowIndices(query._limitOffset, *res)) {
```

Comment:
Maybe the `ExportQueryExecutionTrees` module would be a natural place for all this functionality (it is not really server-related).

Reply:
I agree that this code should probably be moved out of `Server.cpp`. It is the Update equivalent to `sendStreamableResponse` (or even lower down) which is also not in Server. But `ExportQueryExecutionTrees` imo is not a suitable place, because that's not what the function does. Maybe into `ExecuteUpdate` instead?

Reply:
Yes, I like the idea of a new translation unit (.h/.cpp file pair) for the UPDATE stuff, just move it there. The name ` ExecuteUpdate`  (for a class) or ` executeUpdate`  (for a namespace) sound fine.